### PR TITLE
Add Windows shutdown handler

### PR DESCRIPTION
**Pull request by ken2812221**
Exit properly when clicked the red X of Windows Console

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -307,6 +308,14 @@ static void HandleSIGHUP(int)
 {
     fReopenDebugLog = true;
 }
+#else
+static BOOL WINAPI consoleCtrlHandler(DWORD dwCtrlType)
+{
+    fRequestShutdown = true;
+    Sleep(-1);
```
Comment (laanwj) [inline src/init.cpp]: What does sleeping a negative amount of time do?

Comment (ken2812221) [reply src/init.cpp]: For sleeping as long as possible and waiting for main thread quit.

Comment (laanwj) [reply src/init.cpp]: I've checked MSDN docs: https://msdn.microsoft.com/en-us/library/windows/desktop/ms686298(v=vs.85).aspx  and the parameter is a DWORD (unsigned) to -1 is not a valid value.
It seems that there is a constant `INFINITE` to wait infinitely, if that is what is the intent?

Comment (ken2812221) [reply src/init.cpp]: INFINITE is 0xFFFFFFFF, so it might be the same. But to be safe, I would change it to the micro.

**Comment by laanwj** [comment]
utACK ddebde71efe17eb3b1367948e9b728f6a86696ab

**Comment by fanquake** [comment]
Will test shortly.

**Comment by fanquake** [comment]
@jonasschnelli Could you spin up a Windows build for this PR?

**Comment by jonasschnelli** [comment]
Thanks. I think this was long overdue.
utACK ddebde71efe17eb3b1367948e9b728f6a86696ab

Gitian Build: https://bitcoin.jonasschnelli.ch/build/591

**Review by donaloconnor — APPROVED** [review]
tACK ddebde7

I had this change also but didn't get around to submitting it :-)

This works generally but CTRL_CLOSE_EVENT times out after 5 seconds on my OS (Win10) if the main thread hasn't ended gracefully. At that point all threads are terminated.

**Comment by laanwj** [comment]
> This works generally but CTRL_CLOSE_EVENT times out after 5 seconds on my OS (Win10) if the main thread hasn't ended gracefully. At that point all threads are terminated.

I guess there's nothing to be done against that, and it's the same as on UNIX. Every OS has a timeout after which it loses patience and really terminates a process during shutdown.

**Comment by donaloconnor** [comment]
> I guess there's nothing to be done against that, and it's the same as on UNIX. Every OS has a timeout after which it loses patience and really terminates a process during shutdown

Indeed! Just putting there for FYI really.

**Comment by fanquake** [comment]
Added to #13455 for backport.

**Pull request merged**